design/json-substitution-action-design.md

```diff
@@ -0,0 +1,122 @@
+
+Example of Reference to configmap in RestoreSpec
+```yaml
+apiVersion: velero.io/v1
+kind: Restore
+metadata:
+  name: restore-1
+spec:
+    jsonSubstitution:
+      refType: Configmap
+      ref: backup01
+```
+
+### ConfigMap Structure
+- User first needs to provide details on which resources the JSON Substitutions need to be applied. 
+    - For this the user will provide 4 inputs - Nameespaces(for NS Scoped resources), Group, Kind and Name Regex(optional).
+    - If the user does not provide the Name, the JSON Substitutions will be applied to all the resources of the given Group and Kind under the given namespaces.
+
+- Further the use will specify the JSON Substitutions using the structure of kubectl's "JSON Path" based inputs.
+    - Reference: [https://kubernetes.io/docs/reference/kubectl/jsonpath/](https://kubernetes.io/docs/reference/kubectl/jsonpath/)
+    - ![image](https://user-images.githubusercontent.com/36476228/219611516-7fa9527d-c592-4471-8e3e-39ad5e7606db.png)
+    - This filter can be used through golang Package: "k8s.io/client-go/util/jsonpath"
+
+- Last the user will provide the current value regex, basically at the path specified above, we check if the provided regex matches the value at that path. If it does, we substitute the value with the new value provided by the user.
+
+- Sample ConfigMap
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+    name: json-substitution-configmap
+    namespace: velero
+data:
+    json-substitution-configmap.yaml: |
+        resources:
+        - namespaces:
+            - default
+        group: storage.k8s.io
+        kind: StorageClass
+        nameRegex: "disk*"
+        jsonSubstitutions:
+        - path: {..parameters.diskEncryptionSetID}
+            currentValueRegex: ".*"
+            newValue: "new-des"
+        - namespaces:
+            - default
+        group: apps
+        kind: Deployment
+        nameRegex: "nginx*"
+        jsonSubstitutions:
+        - path: spec.template.spec.nodeSelector."kubernetes\.io/hostname"
+            currentValueRegex: ".*"
+            newValue: "new-node"
+```
+## Alternatives Considered
+
+1. JSON Patch
+    - Not very user friendly and too technical for the end user.
+2. JSON Path
+3. RestoreItemAction for each kind of standard substitution
+    - Not an extensible design. If a new kind of substitution is required, a new RestoreItemAction needs to be written.
+4. RIA for JSON Subtitution: The approach of doing JSON Substitution through a RestoreItemAction plugin was considered. But it is likely to have performance implications as the plugin will be invoked for all the resources.
+
+## Security Considerations
+No security impact.
+
+## Compatibility
+Compatibility with existing StorageClass mapping RestoreItemAction and similar plugins needs to be evaluated. 
+
+## Implementation
```

Comment:
Will we also want to add a param to `velero restore` command?

Reply:
Let me check how it is added for resourceFilters configMapRef the CLI is updated.
it will be similar to that.
I will callout in design as well.

Reply:
velero restore create --from-backup backup-1 --resource-modifier-configmap resourcemodifierconfigmap

added the CLI command in the design doc.